server/src/test/java/org/apache/kafka/server/KRaftClusterTest.java

```diff
@@ -262,6 +285,375 @@ private void assertFoobarValue(KafkaClusterTestKit cluster, int expected) throws
+    }
+
+    @FunctionalInterface
+    private interface ThrowingConsumer<T> {
+        void accept(T t) throws Exception;
+    }
+
+    private void waitForTopicListing(Admin admin, List<String> expectedPresent, List<String> expectedAbsent)
+        throws InterruptedException {
+        Set<String> topicsNotFound = new HashSet<>(expectedPresent);
+        Set<String> extraTopics = new HashSet<>();
+        TestUtils.waitForCondition(() -> {
+            Set<String> topicNames = admin.listTopics().names().get();
+            topicsNotFound.removeAll(topicNames);
+            extraTopics.clear();
+            extraTopics.addAll(topicNames.stream().filter(expectedAbsent::contains).collect(Collectors.toSet()));
+            return topicsNotFound.isEmpty() && extraTopics.isEmpty();
+        }, String.format("Failed to find topic(s): %s and NOT find topic(s): %s", topicsNotFound, extraTopics));
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void testUnregisterBroker(boolean usingBootstrapController) throws Exception {
+        try (KafkaClusterTestKit cluster = new KafkaClusterTestKit.Builder(
+            new TestKitNodes.Builder()
+                .setNumBrokerNodes(4)
+                .setNumControllerNodes(3)
+                .build()).build()) {
+            cluster.format();
+            cluster.startup();
+            cluster.waitForReadyBrokers();
+            TestUtils.waitForCondition(() -> brokerIsUnfenced(clusterImage(cluster, 1), 0),
+                "Timed out waiting for broker 0 to be unfenced.");
+            cluster.brokers().get(0).shutdown();
+            TestUtils.waitForCondition(() -> !brokerIsUnfenced(clusterImage(cluster, 1), 0),
+                "Timed out waiting for broker 0 to be fenced.");
+
+            try (Admin admin = createAdminClient(cluster, usingBootstrapController)) {
+                admin.unregisterBroker(0);
+            }
+
+            TestUtils.waitForCondition(() -> brokerIsAbsent(clusterImage(cluster, 1), 0),
+                "Timed out waiting for broker 0 to be fenced.");
+        }
+    }
+
+    private ClusterImage clusterImage(KafkaClusterTestKit cluster, int brokerId) {
+        return cluster.brokers().get(brokerId).metadataCache().currentImage().cluster();
+    }
+
+    private boolean brokerIsUnfenced(ClusterImage image, int brokerId) {
+        BrokerRegistration registration = image.brokers().get(brokerId);
+        if (registration == null) {
+            return false;
+        }
+        return !registration.fenced();
+    }
+
+    private boolean brokerIsAbsent(ClusterImage image, int brokerId) {
+        return !image.brokers().containsKey(brokerId);
+    }
+
+    private Admin createAdminClient(KafkaClusterTestKit cluster, boolean bootstrapController) {
+        Properties props;
+        if (bootstrapController) {
+            props = cluster.newClientPropertiesBuilder().setUsingBootstrapControllers(true).build();
```

Comment:
this call chain is somewhat redundant. I suggest we make it consistent with `ClusterInstance#admin`